draft-ietf-quic-qcram.md

```diff
@@ -127,23 +127,119 @@ considered blocked by the decoder and can not be processed until all entries in
 the range `[1, Depends Index]` have been added.  While blocked, header
 field data MUST remain in the blocked stream's flow control window.
 
-# HPACK extensions
+# Wire Format
+
+QCRAM instructions occur on three stream types, each of which uses a separate
+instruction space:
+
+ - Table updates are carried by HEADERS frames on the control stream, as defined
+   by {{QUIC-HTTP}}.  Frames on this stream modify the dynamic table state
+   without generating output to any particular request.
+ - Acknowledgement of header frame processing is carried by HEADER_ACK frames,
+   running from decoder to encoder.
+ - Finally, the contents of HEADERS and PUSH_PROMISE frames on request streams
+   reference the QPACK table state.
+
+This section describes the instructions which are possible on each stream type.
+
+In order to ensure table consistency, all modifications of the header table
+occur on the control stream rather than on request streams. Request streams
+contain only indexed and literal header entries.
+
+## HEADERS Frames on the Control Stream
+
+### Insert
+
+An addition to the header table starts with the '1' one-bit pattern. If the
+header field name matches the header field name of an entry stored in the static
+table or the dynamic table, the header field name can be represented using the
+index of that entry. In this case, the `S` bit indicates whether the reference
+is to the static (S=1) or dynamic (S=0) table and the index of the entry is
+represented as an integer with an 7-bit prefix (see Section 5.1 of [RFC7541]).
+This value is always non-zero.
+
+~~~~~~~~~~ drawing
+     0   1   2   3   4   5   6   7
+   +---+---+---+---+---+---+---+---+
+   | 1 | S |    Name Index (6+)    |
+   +---+---+-----------------------+
+   | H |     Value Length (7+)     |
+   +---+---------------------------+
+   | Value String (Length octets)  |
+   +-------------------------------+
+~~~~~~~~~~
+{: title="Insert Header Field -- Indexed Name"}
+
+Otherwise, the header field name is represented as a string literal (see Section
+5.2 of [RFC7541]). A value 0 is used in place of the table reference, followed
+by the header field name.
+
+~~~~~~~~~~ drawing
+     0   1   2   3   4   5   6   7
+   +---+---+---+---+---+---+---+---+
+   | 1 |             0             |
```

Comment:
Side-thought sparked by your comment.  There are effectively four instructions:

1. Insert with static table name reference
2. Insert with dynamic table name reference
3. Insert with no reference
4. Duplicate (a.k.a. insert with name+value reference)

We're effectively using two bit prefixes for (1)/(2-3), a sentinel value in the index to differentiate (2)/(3), and (4) gets a single-bit instruction code even though it's the least common.

We could save a few bytes by using two-bit codes for everything and dropping the sentinel zeros for (3), except that HPACK defines string literals to begin on a byte boundary.

Reply:
Having a string literal with a 6-bit prefix would be a little bit annoying to implement, but no more than 15 minutes of hacking (to add the extra argument).  And it would save a whole byte for most, if not all insertions that need a new name.  Names longer than 62 octets are - if not completely unknown - very rare.  I like the idea.